proc/types.go

```diff
@@ -118,22 +119,61 @@ func (dbp *Process) loadPackageMap() error {
 	return nil
 }
 
-func (dbp *Process) loadTypeMap(wg *sync.WaitGroup) {
+type sortFunctionsDebugInfoByLowpc []functionDebugInfo
+
+func (v sortFunctionsDebugInfoByLowpc) Len() int           { return len(v) }
+func (v sortFunctionsDebugInfoByLowpc) Less(i, j int) bool { return v[i].lowpc < v[j].lowpc }
+func (v sortFunctionsDebugInfoByLowpc) Swap(i, j int) {
+	temp := v[i]
+	v[i] = v[j]
+	v[j] = temp
+}
+
+func (dbp *Process) loadDebugInfoMaps(wg *sync.WaitGroup) {
 	defer wg.Done()
 	dbp.types = make(map[string]dwarf.Offset)
+	dbp.functions = []functionDebugInfo{}
 	reader := dbp.DwarfReader()
-	for entry, err := reader.NextType(); entry != nil; entry, err = reader.NextType() {
+	for entry, err := reader.Next(); entry != nil; entry, err = reader.Next() {
 		if err != nil {
 			break
 		}
-		name, ok := entry.Val(dwarf.AttrName).(string)
-		if !ok {
-			continue
+		switch entry.Tag {
+		case dwarf.TagArrayType, dwarf.TagBaseType, dwarf.TagClassType, dwarf.TagStructType, dwarf.TagUnionType, dwarf.TagConstType, dwarf.TagVolatileType, dwarf.TagRestrictType, dwarf.TagEnumerationType, dwarf.TagPointerType, dwarf.TagSubroutineType, dwarf.TagTypedef, dwarf.TagUnspecifiedType:
+			name, ok := entry.Val(dwarf.AttrName).(string)
+			if !ok {
+				continue
+			}
+			if _, exists := dbp.types[name]; !exists {
+				dbp.types[name] = entry.Offset
+			}
+		case dwarf.TagSubprogram:
+			lowpc, ok := entry.Val(dwarf.AttrLowpc).(uint64)
+			if !ok {
+				continue
+			}
+			highpc, ok := entry.Val(dwarf.AttrHighpc).(uint64)
+			if !ok {
+				continue
+			}
+			dbp.functions = append(dbp.functions, functionDebugInfo{lowpc, highpc, entry.Offset})
 		}
-		if _, exists := dbp.types[name]; !exists {
-			dbp.types[name] = entry.Offset
+	}
+	sort.Sort(sortFunctionsDebugInfoByLowpc(dbp.functions))
+}
+
+func (dbp *Process) findFunctionDebugInfo(pc uint64) (dwarf.Offset, error) {
+	i := sort.Search(len(dbp.functions), func(i int) bool {
+		fn := dbp.functions[i]
+		return pc <= fn.lowpc || (fn.lowpc <= pc && pc < fn.highpc)
```

Comment:
The logic here seems wrong, wouldn't we want `pc >= fn.lowpc`? E.g. we want to return `true` if the `pc` we're looking up is within the bounds of `(lowpc...highpc)`, and this initial branch (before the OR) would return true if the argument `pc` is less than `fn.lowpc` -- wouldn't that cause a shortcircuit and give us back the wrong function potentially? Why not remove that first condition and use the range check?